proofs/erc20.md

```diff
@@ -0,0 +1,1030 @@
+
+The detailed mechanism of calculating the location varies by compilers. In Viper, for example, `map[key1][key2]` is stored at the location:
+
+```
+hash(hash(pos(map)) + key1) + key2
+```
+
+where `pos(map)` is the position of `map` appearing in the program, and `+` is the addition modulo `2^256`, while in Solidity, it is stored at:
+
+```
+hash(key2 ++ hash(key1 ++ pos(map)))
+```
+
+where `++` is the byte-array concatenation.
+
+The hashed location abstraction allows to uniformly specify the locations in a form parameterized by the underlying compilers. For example, the location of `map[key1][key2]` can be specified as follows, where `{lang}` is a place-holder to be replaced by the name of the compiler.
+
+```
+#hashedLocation({lang}, pos(map), key1 key2)
+```
+
+This abstraction makes the specification independent of the underlying compilers, enabling it to be reused for differently compiled programs.
+
+
+### Gas Consumption
+
+Regarding the gas consumption of the program, the maximum gas amount is specified to ensure that the program does not consume more gas than the limit.
+
+The verifier proves that the actual gas consumption is less than the provided limit, and also reports the exact amount of gas consumed during the execution. Indeed, it reports a set of the amounts since the gas consumption varies depending on the context (i.e., the input parameter values and the state of the storage). This provides the most precise gas analysis.
+
+
+
+We present the specification for each function defined in the ERC20 standard. Here we focus on explaining the EVM-specific detailed behaviors, referring to the ERC20-K specification for the high-level logic.
+
+
+totalSupply
+-----------
+
+```{.k .erc20-spec .hkg-spec}
+[totalSupply]
+```
+
+`<k>` cell specifies that the execution eventually reaches RETURN instruction meaning that the program will successfully terminate. The RETURN instruction says that a 32-byte return value will be stored in the memory at the location RET_ADDR.
+
+```{.k .erc20-spec .hkg-spec}
+k: #execute => (RETURN RET_ADDR:Int 32 ~> _)
+
+callData: #abiCallData("totalSupply", .TypedArgs)
+```
+
+`<localMem>` cell specifies that the local memory is empty in the beginning, but in the end, it will contain some contents including the return value `TOTAL`, the total supply of the tokens.
+
+The other entries are represented by the wildcard symbol `_`, meaning that they are not relevant to the functional correctness of the program.
+
+```{.k .erc20-spec .hkg-spec}
+localMem:
+      .Map
+    =>
+    (
+      .Map[ RET_ADDR := #asByteStackInWidth(TOTAL, 32) ]
+      _:Map
+    )
+```
+
+`<gas>` cell specifies the maximum gas amount. Here we give a loose upper-bound for the demonstration purpose. In practice, one should set a reasonable amount of the gas limit to see if the program does not consume too much gas (i.e., no gas leakage).
```

Comment:
`the demonstration purpose` => `demonstration purposes`.

Reply:
So we are not verifying the exact gas consumption?

Reply:
Say what `{gascap}` is and where it is defined.

Reply:
Specifying the *exact* gas consumption is fragile. Note that the different compiler version will affect it depending on their optimizations. Also, note that the exact gas consumption is indeed a *set* of the gas consumptions, since the gas price for each instruction varies by the current state of the storage.

> So we are not verifying the exact gas consumption?

Reply:
Interesting ... we should say that somewhere, because people will be interested in gas consumption.
Also, once the dust settles, we should have a proof where we also calculate the exact gas consumption, symbolically.  Are we going to have the same complexity in IELE, where the gas price of each instruction varies by the current state of the storage?